core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java

```diff
@@ -2467,6 +2506,367 @@ public Type visitGroupingOperation(GroupingOperation node, StackableAstVisitorCo
+                            !isDateTimeType(parameterType) &&
+                            !parameterType.equals(BOOLEAN)) {
+                        coerceType(parameter, parameterType, VARCHAR, "JSON path parameter");
+                        passedType = VARCHAR;
+                    }
+                    else {
+                        passedType = parameterType;
+                    }
+                }
+
+                types.put(parameterName, passedType);
+                fields.add(new RowType.Field(Optional.of(parameterName), passedType));
+            }
+
+            Type parametersRowType = DUMMY_ROW_TYPE;
+            if (!pathParameters.isEmpty()) {
+                parametersRowType = RowType.from(fields.build());
+            }
+
+            // ANALYZE JSON PATH
+            Map<String, Type> typesMap = types.buildOrThrow();
+            JsonPathAnalysis pathAnalysis = new JsonPathAnalyzer(
+                    plannerContext.getMetadata(),
+                    session,
+                    createConstantAnalyzer(plannerContext, accessControl, session, ExpressionAnalyzer.this.parameters, WarningCollector.NOOP))
+                    .analyzeJsonPath(jsonPathInvocation.getJsonPath(), typesMap);
+            jsonPathAnalyses.put(NodeRef.of(node), pathAnalysis);
+
+            return ImmutableList.of(
+                    JSON_2016, // input expression
+                    plannerContext.getTypeManager().getType(TypeId.of(JsonPath2016Type.NAME)), // parsed JSON path representation
+                    parametersRowType); // passed parameters
+        }
+
+        private ResolvedFunction getInputFunction(Type type, JsonFormat format, Node node)
+        {
+            QualifiedName name;
+            switch (format) {
+                case JSON:
+                    if (UNKNOWN.equals(type) || isCharacterStringType(type)) {
+                        name = QualifiedName.of("$varchar_to_json");
+                    }
+                    else if (isStringType(type)) {
+                        name = QualifiedName.of("$varbinary_to_json");
+                    }
+                    else {
+                        throw semanticException(TYPE_MISMATCH, node, format("Cannot read input of type %s as JSON using formatting %s", type, format));
+                    }
+                    break;
+                case UTF8:
+                    name = QualifiedName.of("$varbinary_utf8_to_json");
+                    break;
+                case UTF16:
+                    name = QualifiedName.of("$varbinary_utf16_to_json");
+                    break;
+                case UTF32:
+                    name = QualifiedName.of("$varbinary_utf32_to_json");
+                    break;
+                default:
+                    throw new UnsupportedOperationException("Unexpected format: " + format);
+            }
+            try {
+                return plannerContext.getMetadata().resolveFunction(session, name, fromTypes(type, BOOLEAN));
```

Comment:
does `resolveFunction` already assumes that coercion may be needed on arguments?

Reply:
Yes, the `resolvedFunction` might require other argument types than the actual passed argument types. That's why it's followed by coercion of arguments.